### PR TITLE
BUG: Tolerant intersection

### PR DESCRIPTION
**Pull request by cpelley**
This PR stops intersection from wrapping grids for the following situation:

The grid is being split in two at either edge due to the point being a value of (base + period).

This is not at all a neat solution but it yield the desired result.  Any alternate solutions are welcome.

Further context in #1662

### CODE REVIEW

**Comment by cpelley** [comment]
Not sure why the travis build failed.

**Comment by pelson** [comment]
Thanks @cpelley - looks like you've got a solution which fits all of the existing tests and fits your requirements, nicely done!
I'm very conscious that this has added a *lot* of complexity to the determination of subsets, which essentially means that this will be the last time that anybody can fully grasp the subset computation comfortably in their head. For that reason, I think this needs a whole lot more commenting, documentation of the overall *aim* of subsets, and I would suggest splitting some of the indexing lines into sensibly named (aka self documenting) variable definitions.
Because of the complexity, I currently cannot be certain that you have tested all of the cases that you have added, so I will defer reviewing the tests until the logic is a little easier to review.

:+1: in principle on finding a solution to the problem, and I'm happy to help moving it forwards with this PR.

**Comment by cpelley** [comment]
Thanks for your comment @pelson.  I'll make changes based on your comment ASAP.

**Comment by cpelley** [comment]
Done, cheers @pelson, ready for review.

**Comment by cpelley** [comment]
ping :)

**Comment by cpelley** [comment]
ping x2
Is there anything stopping this PR from getting a look into?

File: lib/iris/cube.py [hunk 1]
```diff
@@ -2278,8 +2282,68 @@ def _intersect_modulus(self, coord, minimum, maximum, min_inclusive,
...
+                                             inside_indices[-1] + 1)]
+                        # Unwrap points
+                        elif np.isclose(coord.points[0],
+                                        coord.points[-1] + modulus):
+                            points[0] = coord.points[0]
+                            subsets = [slice(inside_indices[0],
+                                             inside_indices[-1] + 1)]
+                else:
+                    subsets = [
+                        slice(index_of_second_chunk, None),
```
Comment (rhattersley) [inline lib/iris/cube.py]: This line changed during your development but doesn't seem to have been any corresponding change in the tests - perhaps it hasn't been covered by the tests? Or perhaps the tests didn't always pass?

File: lib/iris/cube.py [hunk 2]
```diff
@@ -2278,8 +2282,68 @@ def _intersect_modulus(self, coord, minimum, maximum, min_inclusive,
...
+                            bounds[-1, :] = coord.bounds[-1, :]
+                            points[-1] = coord.points[-1]
+                            subsets = [slice(inside_indices[0],
+                                             inside_indices[-1] + 1)]
+                        # Unwrap points
+                        elif np.isclose(coord.points[-1],
+                                        coord.points[0] + modulus):
+                            points[-1] = coord.points[-1]
+                            subsets = [slice(inside_indices[0],
+                                       inside_indices[-1] + 1)]
```
Comment (rhattersley) [inline lib/iris/cube.py]: This code path isn't tested.

File: lib/iris/cube.py [hunk 3]
```diff
@@ -2278,8 +2282,68 @@ def _intersect_modulus(self, coord, minimum, maximum, min_inclusive,
...
+                            bounds[0, :] = coord.bounds[0, :]
+                            points[0] = coord.points[0]
+                            subsets = [slice(inside_indices[0],
+                                             inside_indices[-1] + 1)]
+                        # Unwrap points
+                        elif np.isclose(coord.points[0],
+                                        coord.points[-1] + modulus):
+                            points[0] = coord.points[0]
+                            subsets = [slice(inside_indices[0],
+                                             inside_indices[-1] + 1)]
```
Comment (rhattersley) [inline lib/iris/cube.py]: This code path isn't tested.

File: lib/iris/cube.py (hunk 1 above)
Comment (rhattersley) [inline lib/iris/cube.py]: NB. This code path *is* tested so the tests probably broke at some point.

**Comment by rhattersley** [comment]
This is still a complex, hard-to-follow change so a lot of my confidence in this change is going to come from the testing. (Hence the comments about untested code paths.)

It'd be great if `Cube._intersect_modulus` could be re-factored to break it down into "bite-size" functions... but if you don't have time for that then I'm OK with merging as it stands.

**Comment by rhattersley** [comment]
> It'd be great if Cube._intersect_modulus could be re-factored

For example:
```
$ flake8 --max-complexity=10 lib/iris/cube.py
...
lib/iris/cube.py:2204:1: C901 'Cube._intersect_modulus' is too complex (17)
...
```

File: lib/iris/cube.py (hunk 2 above)
Comment (cpelley) [inline lib/iris/cube.py]: Thanks @rhattersley

File: lib/iris/cube.py (hunk 3 above)
Comment (cpelley) [inline lib/iris/cube.py]: Thank you @rhattersley

**Comment by cpelley** [comment]
Thank you @rhattersley for your time on reviewing this.  I totally agree about the complexity, hence my description in the PR.  I'll certainly ensure complete test coverage as you have quite rightly suggested.  With regards to refactoring, I'll have a look to see if there are quick refactors possible without too much controversy.  I'll open a new ticket to capture this requirement if I don't think the refactor is complete (it certainly needs it as I found it difficult enough to follow the logic before I introduced my changes, let alone now).

`flake8` - Nice, I didn't know about this tool.  I'll read up about what heuristics it uses to determine complexity.  Sounds VERY useful.

Cheers @rhattersley, I'll get back to you ASAP

File: lib/iris/cube.py (hunk 2 above)
Comment (cpelley) [inline lib/iris/cube.py]: I managed to write a test for this that was not actually testing it - I have corrected this now.
Out of interest how did you spot this?  Did you but break points here like I just did?

**Comment by cpelley** [comment]
I have added a test and modified an existing one so now both paths are tested.

Now looking at the possibility of a refactor, will let you know if I get anywhere ASAP

**Comment by cpelley** [comment]
I have a made I think a pretty good refactor (makes a lot more sense now).

Over to you, cheers @rhattersley

**Comment by cpelley** [comment]
Apologies, forgot to push back up.  Ready now.

**Comment by cpelley** [comment]
You happy with the refactor @rhattersley ?

I think intersection should eventually be pulled out into its own module but I have approach the refactor from the perspective of increasing code clarity, simplification and removing duplication.  I think this approach is most important right now.

File: lib/iris/cube.py [hunk 4]
```diff
@@ -2197,6 +2201,90 @@ def create_coords(src_coords, add_coord):
...
+        delta = coord.points[inside_indices] - points[inside_indices]
+        step = np.rint(np.diff(delta) / modulus)
+        non_zero_step_indices = np.nonzero(step)[0]
+
+        def dim_coord_subset():
+            """
+            Derive the subset for dimension coordinates.
+
+            Ensure that we do not wrap if blocks are at the very edge.  That
+            is, if the very edge if wrapped and corresponds to base + period,
```
Comment (rhattersley) [inline lib/iris/cube.py]: Typo: "edge **if** wrapped" -> "edge **is** wrapped"

**Comment by rhattersley** [comment]
Thanks for the refactor. It's a definite improvement.

If you fix the typo and squash the commits I'm :+1: to merge.

**Comment by rhattersley** [comment]
Oops. Just realised it could do with a [what's new entry](https://github.com/SciTools/iris/tree/master/docs/iris/src/whatsnew/contributions_1.9). Feel free to leave it as a separate commit - whatever's easier.

**Comment by cpelley** [comment]
Done, cheers @rhattersley

**Pull request merged**